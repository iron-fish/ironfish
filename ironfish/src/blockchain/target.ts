/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import type Serde from '../serde'

function max(a: BigInt, b: BigInt): BigInt {
  if (a > b) {
    return a
  } else {
    return b
  }
}

/**
 * Courtesy of https://coolaj86.com/articles/convert-js-bigints-to-typedarrays/
 *
 * Convert a Buffer to a big integer number, in big endian format.
 *
 * I'm concerned about efficiency here. Converting a string and back and... WTF?
 * Every block hash attempt has to be converted to a Target, so this is a function
 * that should be optimized. We may want to compile this to wasm if there isn't
 * a less janky way to do it.
 *
 * I'm pushing it out like this for now so I can focus on bigger architecture concerns.
 *
 * Sorry.
 */
export function bytesToBigInt(bytes: Buffer): bigint {
  const hex: string[] = []
  if (bytes.length === 0) {
    return BigInt(0)
  }
  bytes.forEach(function (i) {
    let h = i.toString(16)
    if (h.length % 2) {
      h = '0' + h
    }
    hex.push(h)
  })

  return BigInt('0x' + hex.join(''))
}

export function bigIntToBytes(bigint: bigint): Buffer {
  let hex = bigint.toString(16)
  if (hex.length % 2) {
    hex = '0' + hex
  }

  const len = hex.length / 2
  const u8 = Buffer.alloc(len)

  let i = 0
  let j = 0
  while (i < len) {
    u8[i] = parseInt(hex.slice(j, j + 2), 16)
    i += 1
    j += 2
  }

  return u8
}

/**
 * The bound divisor of the difficulty, used to update difficulty (and subsequently target).
 * We are taking in large part Ethereum's dynamic difficulty calculation,
 * with the exeption of 'uncles' and 'difficulty bomb' as a concept
 * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
 * original algorithm:
 * diff = (parent_diff +
 *         (parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) // 10, -99))
 *        ) + 2**((current_block_number // 100000) — 2)
 * Note we are not including the difficulty bomb (which is this part: 2**((current_block_number // 100000) — 2))
 * So the algorithm for target is:
 * diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
 * note that timestamps above are in seconds, and JS timestamps are in ms
 * The bound divisor of the difficulty is the '2048' part of that equation
 */
const DIFFICULTY_ADJUSTMENT_DENOMINATOR = 2048

/**
 *  Minimum difficulty, which is equivalent to maximum target
 */
const MIN_DIFFICULTY = BigInt(131072)

/**
 * Maximum target, which is equivalent of minimum difficulty of 131072
 * target == 2**256 / difficulty
 */
const MAX_TARGET = BigInt(
  '883423532389192164791648750371459257913741948437809479060803100646309888',
)

/**
 *  Maximum number to represent a 256 bit number, which is 2**256 - 1
 */
const MAX_256_BIT_NUM = BigInt(
  '115792089237316195423570985008687907853269984665640564039457584007913129639935',
)

export class Target {
  targetValue: bigint
  constructor(targetValue: bigint | BigInt | Buffer | string | number | undefined = undefined) {
    if (targetValue === undefined) {
      this.targetValue = BigInt(0)
    } else {
      let candidate
      if (targetValue instanceof Buffer) {
        candidate = bytesToBigInt(targetValue)
      } else {
        candidate = BigInt(targetValue)
      }
      if (candidate > MAX_256_BIT_NUM) {
        throw new Error('Target value exceeds max target')
      } else {
        this.targetValue = candidate
      }
    }
  }

  /**
   * Maximum target (in terms of difficulty), which is equivalent of
   * minimum difficulty of 131072
   * maximum target == minimum difficulty
   * target == 2**256 / difficulty
   */
  static maxTarget(): Target {
    return new Target(MAX_TARGET)
  }

  /**
   * The initial target on the genesis block.
   *
   * This will need to be hand-tuned.
   */
  static initialTarget(): Target {
    return this.maxTarget()
  }

  /**
   * Calculate the target for the current block given the timestamp in that
   * block's header, the pervious block's timestamp and previous block's target.
   *
   * To verify whether a target on a block is correct, pass in the timestamp in its header,
   * its previous block's timestamp, and its previous block's target
   * and compare the resulting target to what is specified on the current block header
   *
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateTarget(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockTarget: Target,
  ): Target {
    const parentDifficulty = previousBlockTarget.toDifficulty()
    const difficulty = Target.calculateDifficulty(
      time,
      previousBlockTimestamp,
      parentDifficulty,
    )

    return Target.fromDifficulty(difficulty)
  }

  /**
   *
   * Calculate the difficulty for the current block given the timestamp in that
   * block's header, the pervious block's timestamp and previous block's target.
   *
   * Note that difficulty == 2**256 / target and target == 2**256 / difficulty
   *
   * Returns the difficulty for a block given it timestamp for that block and its parent.
   * @param time the block's timestamp for which the target is calcualted for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateDifficulty(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockDifficulty: BigInt,
  ): BigInt {
    // We are taking in large part Ethereum's dynamic difficulty calculation,
    // with the exeption of 'uncles' and 'difficulty bomb' as a concept
    // https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
    // original algorithm:
    // diff = (parent_diff +
    //         (parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) // 10, -99))
    //        ) + 2**((current_block_number // 100000) — 2)
    // Note we are not including the difficulty bomb (which is this part: 2**((current_block_number // 100000) — 2))
    // So the algorithm we're taking is:
    // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    // note that timestamps above are in seconds, and JS timestamps are in ms

    // max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    const diffInSeconds = (time.getTime() - previousBlockTimestamp.getTime()) / 1000
    const sign = BigInt(Math.max(1 - Math.floor(diffInSeconds / 10), -99))
    const offset = BigInt(previousBlockDifficulty) / BigInt(DIFFICULTY_ADJUSTMENT_DENOMINATOR)

    // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    const difficulty = max(
      BigInt(previousBlockDifficulty) + offset * sign,
      Target.minDifficulty(),
    )

    return difficulty
  }

  /**
   * Returns the minimum difficulty that can be used for Iron Fish blocks
   * To be used in calculateTarget for easier mocking
   */
  static minDifficulty(): BigInt {
    return MIN_DIFFICULTY
  }

  /**
   * Converts difficulty to Target
   */
  static fromDifficulty(difficulty: BigInt): Target {
    if (difficulty == BigInt(1)) {
      return new Target(MAX_256_BIT_NUM)
    }
    return new Target(BigInt(2) ** BigInt(256) / BigInt(difficulty))
  }

  /**
   * Return the difficulty representation as a big integer
   */
  toDifficulty(): bigint {
    return BigInt(2) ** BigInt(256) / this.targetValue
  }

  /**
   * Add the given amout to the target's value. A negative amount makes the target
   * harder to achieve, a positive one makes it easier.
   *
   * If adjustment would make target negative or higher than max allowed value,
   * the current target is returned unchanged.
   */
  adjust(amount: bigint): Target {
    let adjusted = this.targetValue + amount
    if (adjusted > MAX_TARGET || adjusted < 0) {
      adjusted = this.targetValue
    }
    return new Target(adjusted)
  }

  /**
   * Return whether or not this target meets the requirements of the given target,
   * which is to say, this has a lower numeric value then the provided one.
   */
  static meets(hashValue: BigInt, target: Target): boolean {
    return hashValue <= target.targetValue
  }

  /**
   * Return the target number as a big integer
   */
  asBigInt(): bigint {
    return this.targetValue
  }

  /**
   * Big endian, and since even after 20 years in the industry, I can't remember
   * what that means, the most significant byte is in the 0th index of the array.
   *
   * The resulting byte array is always left padded with 0s to be 32 bytes long
   */
  asBytes(): Buffer {
    const bytes = bigIntToBytes(this.targetValue)
    const result = Buffer.alloc(32)
    result.set(bytes, 32 - bytes.length)
    return result
  }
}

export class TargetSerde implements Serde<Target, string> {
  equals(target1: Target, target2: Target): boolean {
    return target1.targetValue === target2.targetValue
  }

  serialize(target: Target): string {
    return target.targetValue.toString()
  }

  deserialize(data: string): Target {
    if (typeof data !== 'string') {
      throw new Error('Can only deserialize Target from string')
    }
    return new Target(data)
  }
}

export const TargetSerdeInstance = new TargetSerde()
