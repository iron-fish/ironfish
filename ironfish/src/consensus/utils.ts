/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export function isExpiredSequence(expiration: number, sequence: number): boolean {
  return expiration !== 0 && expiration <= sequence
}
