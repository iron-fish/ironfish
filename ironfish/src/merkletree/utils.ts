/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { LeafIndex, NodeIndex } from './merkletree'

/**
 * Is the given leaf a right child or left child of its parent node.
 *
 * Leaves are added in order, so this is the same as asking if the index
 * is an od number
 */
export function isRight(index: LeafIndex): boolean {
  return index % 2 === 1
}

/**
 * Is the given node index the empty node above the root node?
 */
export function isEmpty(index: NodeIndex | undefined): index is undefined | 0 {
  return index === 0 || index === undefined
}

/**
 * The depth of the tree when it contains a certain number of leaf nodes
 */
export function depthAtLeafCount(size: number): number {
  if (size === 0) {
    return 0
  }

  if (size === 1) {
    return 1
  }

  return Math.floor(Math.log2(size - 1)) + 2
}
