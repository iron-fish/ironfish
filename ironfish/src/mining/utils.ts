/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import bufio from 'bufio'
import { SerializedBlockTemplate } from '../serde/BlockTemplateSerde'

const MINEABLE_BLOCK_HEADER_SIZE = 180
export const MINEABLE_BLOCK_HEADER_GRAFFITI_OFFSET = MINEABLE_BLOCK_HEADER_SIZE - 32

export function mineableHeaderString(header: SerializedBlockTemplate['header']): Buffer {
  const bw = bufio.write(MINEABLE_BLOCK_HEADER_SIZE)
  bw.writeBytes(Buffer.from(header.randomness, 'hex'))
  bw.writeU32(header.sequence)
  bw.writeHash(header.previousBlockHash)
  bw.writeHash(header.noteCommitment)
  bw.writeHash(Buffer.from(header.transactionCommitment, 'hex'))
  bw.writeHash(header.target)
  bw.writeU64(header.timestamp)
  bw.writeBytes(Buffer.from(header.graffiti, 'hex'))
  return bw.render()
}

// deserialize into a partial header
export function minedPartialHeader(data: Buffer): SerializedBlockTemplate['header'] {
  const br = bufio.read(data)
  const randomness = br.readBytes(8)
  const sequence = br.readU32()
  const previousBlockHash = br.readHash()
  const noteCommitment = br.readHash()
  const transactionCommitment = br.readHash()
  const target = br.readBytes(32)
  const timestamp = br.readU64()
  const graffiti = br.readBytes(32)

  return {
    randomness: randomness.toString('hex'),
    sequence: sequence,
    previousBlockHash: previousBlockHash.toString('hex'),
    target: target.toString('hex'),
    timestamp: timestamp,
    graffiti: graffiti.toString('hex'),
    noteCommitment: noteCommitment.toString('hex'),
    transactionCommitment: transactionCommitment.toString('hex'),
  }
}
