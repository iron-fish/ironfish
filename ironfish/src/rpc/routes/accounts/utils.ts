/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Account } from '../../../account'
import { IronfishNode } from '../../../node'
import { ValidationError } from '../../adapters'

export function getAccount(node: IronfishNode, name?: string): Account {
  if (name) {
    const account = node.accounts.getAccountByName(name)
    if (account) {
      return account
    }
    throw new ValidationError(`No account with name ${name}`)
  }

  const defaultAccount = node.accounts.getDefaultAccount()
  if (defaultAccount) {
    return defaultAccount
  }

  throw new ValidationError(
    `No account is currently active.\n\n` +
      `Use ironfish accounts:create <name> to first create an account`,
  )
}
