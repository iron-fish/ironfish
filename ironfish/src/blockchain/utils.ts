/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Transaction } from '../primitives/transaction'
import { JsonSerializable } from '../serde'
import { IDatabaseTransaction } from '../storage'
import { Blockchain } from './blockchain'

/**
 * Sanity check to check that heaviest head exists, and trees match it
 * If we just added a block that puts trees in a bad state, abort it
 * as its incorrect
 */
export async function checkTreeMatchesHeaviest<
  E,
  H,
  T extends Transaction<E, H>,
  SE extends JsonSerializable,
  SH extends JsonSerializable,
  ST
>(chain: Blockchain<E, H, T, SE, SH, ST>, tx?: IDatabaseTransaction): Promise<boolean> {
  const noteRoot = await chain.notes.rootHash(tx)
  const nullifierRoot = await chain.nullifiers.rootHash(tx)

  const heaviestHead = chain.head
  if (!heaviestHead) {
    chain.logger.error(`No heaviest head — should never happen`)
    return false
  }

  const heaviestBlock = await chain.getBlock(heaviestHead.hash, tx)
  if (!heaviestBlock) {
    chain.logger.error(`No heaviest block — should never happen`)
    return false
  }

  if (
    !chain.strategy
      .noteHasher()
      .hashSerde()
      .equals(noteRoot, heaviestBlock.header.noteCommitment.commitment)
  ) {
    const blockNoteSize = heaviestBlock.header.noteCommitment.size
    const noteSize = await chain.notes.size(tx)

    const noteRootSerialized = chain.strategy.noteHasher().hashSerde().serialize(noteRoot)
    const blockRootSerialized = chain.strategy
      .noteHasher()
      .hashSerde()
      .serialize(heaviestBlock.header.noteCommitment.commitment)

    chain.logger.error(
      `Note Merkle Tree is in a BAD STATE: \n
      Heaviest head is ${heaviestBlock.header.hash.toString('hex')} seq ${
        heaviestBlock.header.sequence
      }
        Note tree size: ${noteSize} \n
        Note root: ${
          noteRootSerialized ? (noteRootSerialized as Buffer).toString('hex') : '???'
        } \n
        Block commitment tree size: ${blockNoteSize}\n
        Block commitment: ${
          blockRootSerialized ? (blockRootSerialized as Buffer).toString('hex') : '???'
        }\n`,
    )

    chain.logger.debug(`TREES IN BAD STATE`)
    return false
  }

  if (
    !chain.strategy
      .nullifierHasher()
      .hashSerde()
      .equals(nullifierRoot, heaviestBlock.header.nullifierCommitment.commitment)
  ) {
    const nullifierSize = await chain.nullifiers.size(tx)
    const blockNullifierSize = heaviestBlock.header.nullifierCommitment.size
    chain.logger.error(
      `After adding block ${heaviestBlock.header.hash.toString('hex')} seq ${
        heaviestBlock.header.sequence
      } Nullifier Merkle Tree is in a BAD STATE: \n
      Nullifier tree size: ${nullifierSize} \n
      Block commitment tree size: ${blockNullifierSize}`,
    )
    chain.logger.debug(`TREES IN BAD STATE`)
    return false
  }

  return true
}
