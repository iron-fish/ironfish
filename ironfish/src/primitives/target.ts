/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { BigIntUtils } from '../utils/bigint'

/**
 *  Minimum difficulty, which is equivalent to maximum target
 */
const MIN_DIFFICULTY = 131072n

/**
 * Maximum target, which is equivalent of minimum difficulty of 131072
 * target == 2**256 / difficulty
 */
const MAX_TARGET = 883423532389192164791648750371459257913741948437809479060803100646309888n

/**
 *  Maximum number to represent a 256 bit number, which is 2**256 - 1
 */
const MAX_256_BIT_NUM =
  115792089237316195423570985008687907853269984665640564039457584007913129639935n

export class Target {
  targetValue: bigint
  constructor(targetValue: bigint | Buffer | string | number) {
    const candidate =
      targetValue instanceof Buffer ? BigIntUtils.fromBytesBE(targetValue) : BigInt(targetValue)

    if (candidate > MAX_256_BIT_NUM) {
      throw new Error('Target value exceeds max target')
    } else {
      this.targetValue = candidate
    }
  }

  /**
   * Maximum target (in terms of difficulty), which is equivalent of
   * minimum difficulty of 131072
   * maximum target == minimum difficulty
   * target == 2**256 / difficulty
   */
  static maxTarget(): Target {
    return new Target(MAX_TARGET)
  }

  static minTarget(): Target {
    return new Target(0)
  }

  /**
   * Calculate the target for the current block given the timestamp in that
   * block's header, the previous block's timestamp and previous block's target.
   *
   * To verify whether a target on a block is correct, pass in the timestamp in its header,
   * its previous block's timestamp, and its previous block's target
   * and compare the resulting target to what is specified on the current block header
   *
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateTarget(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockTarget: Target,
    targetBlockTimeInSeconds: number,
    targetBucketTimeInSeconds: number,
  ): Target {
    const parentDifficulty = previousBlockTarget.toDifficulty()

    const difficulty = Target.calculateDifficulty(
      time,
      previousBlockTimestamp,
      parentDifficulty,
      targetBlockTimeInSeconds,
      targetBucketTimeInSeconds,
    )

    return Target.fromDifficulty(difficulty)
  }

  /**
   * Calculate the difficulty for the current block given the timestamp in that
   * block's header, the previous block's timestamp and previous block's target.
   *
   * Note that difficulty == 2**256 / target and target == 2**256 / difficulty
   *
   * Algorithm: difficulty = parentDifficulty - (parentDifficulty / 2048) * bucket
   * Where bucket is how many steps (in TARGET_BUCKET_TIME_IN_SECONDS) the new time is away from
   * our target bucket range, e.g. for target block time of 60 seconds (with +/-5 seconds forgiveness):
   * 35 - 45 seconds: bucket -2
   * 45 - 55 seconds: bucket -1
   * 55 - 65 seconds: bucket 0
   * 65 - 75 seconds: bucket 1
   * 75 - 85 seconds: bucket 2
   * .. and so on
   *
   * Returns the difficulty for a block given it timestamp for that block and its parent.
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateDifficulty(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockDifficulty: bigint,
    targetBlockTimeInSeconds: number,
    targetBucketTimeInSeconds: number,
  ): bigint {
    const diffInSeconds = (time.getTime() - previousBlockTimestamp.getTime()) / 1000

    let bucket = Math.floor(
      (diffInSeconds - targetBlockTimeInSeconds + Math.floor(targetBucketTimeInSeconds / 2)) /
        targetBucketTimeInSeconds,
    )

    // Should not change difficulty by more than 99 buckets from last block's difficulty
    bucket = Math.min(bucket, 99)

    const difficulty =
      previousBlockDifficulty - (previousBlockDifficulty / 2048n) * BigInt(bucket)

    return BigIntUtils.max(difficulty, Target.minDifficulty())
  }

  /**
   * Returns the minimum difficulty that can be used for Iron Fish blocks
   * To be used in calculateTarget for easier mocking
   */
  static minDifficulty(): bigint {
    return MIN_DIFFICULTY
  }

  /**
   * Converts difficulty to Target
   */
  static fromDifficulty(difficulty: bigint): Target {
    if (difficulty <= Target.minDifficulty()) {
      return Target.maxTarget()
    }
    return new Target((2n ** 256n / difficulty).valueOf())
  }

  /**
   * Return the difficulty representation as a big integer
   */
  toDifficulty(): bigint {
    if (this.targetValue <= 1n) {
      return MAX_256_BIT_NUM
    }
    return 2n ** 256n / this.targetValue
  }

  /**
   * Add the given amount to the target's value. A negative amount makes the target
   * harder to achieve, a positive one makes it easier.
   *
   * If adjustment would make target negative or higher than max allowed value,
   * the current target is returned unchanged.
   */
  adjust(amount: bigint): Target {
    let adjusted = this.targetValue + amount
    if (adjusted > MAX_TARGET || adjusted < 0) {
      adjusted = this.targetValue
    }
    return new Target(adjusted)
  }

  /**
   * Return whether or not this target meets the requirements of the given target,
   * which is to say, this has a lower numeric value then the provided one.
   */
  static meets(hashValue: BigInt, target: Target): boolean {
    return hashValue <= target.targetValue
  }

  /**
   * Return the target number as a big integer
   */
  asBigInt(): bigint {
    return this.targetValue
  }

  equals(other: Target): boolean {
    return this.targetValue === other.targetValue
  }
}
