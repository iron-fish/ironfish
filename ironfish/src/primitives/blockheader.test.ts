/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { blake3 } from '@napi-rs/blake-hash'
import { v4 as uuid } from 'uuid'
import { GraffitiUtils } from '../utils'
import {
  BlockHeader,
  BlockHeaderSerde,
  isBlockHeavier,
  isBlockLater,
  NULL_NODE,
  TRANSACTION_ROOT_PERSONALIZATION,
  transactionMerkleRoot,
} from './blockheader'
import { Target } from './target'

const level = (l: number) => Buffer.concat([TRANSACTION_ROOT_PERSONALIZATION, Buffer.from([l])])

describe('transactionMerkleRoot', () => {
  it('calculates merkle root with 0 transactions', () => {
    const root = transactionMerkleRoot([])

    expect(root.equals(blake3(TRANSACTION_ROOT_PERSONALIZATION))).toBe(true)
  })

  it('calculates merkle root with 1 transaction', () => {
    const hashes = [...new Array(1)].map((_) => blake3(uuid()))
    const root = transactionMerkleRoot(hashes)

    const expectedRoot = blake3(Buffer.concat([level(0), hashes[0], NULL_NODE]))
    expect(root.equals(expectedRoot)).toBe(true)
  })

  it('calculates merkle root with 2 transactions', () => {
    const hashes = [...new Array(2)].map((_) => blake3(uuid()))
    const root = transactionMerkleRoot(hashes)

    const expectedRoot = blake3(Buffer.concat([level(0), hashes[0], hashes[1]]))
    expect(root.equals(expectedRoot)).toBe(true)
  })

  it('calculates merkle root with 3 transactions', () => {
    const hashes = [...new Array(3)].map((_) => blake3(uuid()))
    const root = transactionMerkleRoot(hashes)

    const left = blake3(Buffer.concat([level(0), hashes[0], hashes[1]]))
    const right = blake3(Buffer.concat([level(0), hashes[2], NULL_NODE]))

    const expectedRoot = blake3(Buffer.concat([level(1), left, right]))
    expect(root.equals(expectedRoot)).toBe(true)
  })

  it('calculates merkle root with 4 transactions', () => {
    const hashes = [...new Array(3)].map((_) => blake3(uuid()))
    const root = transactionMerkleRoot(hashes)

    const left = blake3(Buffer.concat([level(0), hashes[0], hashes[1]]))
    const right = blake3(Buffer.concat([level(0), hashes[2], hashes[3]]))

    const expectedRoot = blake3(Buffer.concat([level(1), left, right]))
    expect(root.equals(expectedRoot)).toBe(true)
  })

  it('calculates merkle root with 11 transactions', () => {
    const hashes = [...new Array(11)].map((_) => blake3(uuid()))
    const root = transactionMerkleRoot(hashes)

    const l1 = blake3(Buffer.concat([level(0), hashes[0], hashes[1]]))
    const r1 = blake3(Buffer.concat([level(0), hashes[2], hashes[3]]))
    const l1r1 = blake3(Buffer.concat([level(1), l1, r1]))

    const l2 = blake3(Buffer.concat([level(0), hashes[4], hashes[5]]))
    const r2 = blake3(Buffer.concat([level(0), hashes[6], hashes[7]]))
    const l2r2 = blake3(Buffer.concat([level(1), l2, r2]))

    const l3 = blake3(Buffer.concat([level(0), hashes[8], hashes[9]]))
    const r3 = blake3(Buffer.concat([level(0), hashes[10], NULL_NODE]))
    const l3r3 = blake3(Buffer.concat([level(1), l3, r3]))

    const l4 = blake3(Buffer.concat([level(0), NULL_NODE, NULL_NODE]))
    const r4 = blake3(Buffer.concat([level(0), NULL_NODE, NULL_NODE]))
    const l4r4 = blake3(Buffer.concat([level(1), l4, r4]))

    const l1r1l2r2 = blake3(Buffer.concat([level(2), l1r1, l2r2]))
    const l3r3l4r4 = blake3(Buffer.concat([level(2), l3r3, l4r4]))

    const expectedRoot = blake3(Buffer.concat([level(3), l1r1l2r2, l3r3l4r4]))
    expect(root.equals(expectedRoot)).toBe(true)
  })
})

describe('BlockHeader', () => {
  it('checks equal block headers', () => {
    const header1 = new BlockHeader(
      5,
      Buffer.alloc(32),
      Buffer.alloc(32, 'header'),
      Buffer.alloc(32, 'transactionRoot'),
      new Target(17),
      BigInt(25),
      new Date(1598467858637),
      Buffer.alloc(32),
    )

    const header2 = new BlockHeader(
      5,
      Buffer.alloc(32),
      Buffer.alloc(32, 'header'),
      Buffer.alloc(32, 'transactionRoot'),
      new Target(17),
      BigInt(25),
      new Date(1598467858637),
      Buffer.alloc(32),
    )

    expect(header1.equals(header2)).toBe(true)

    // sequence
    header2.sequence = 6
    expect(header1.equals(header2)).toBe(false)
    header2.sequence = header1.sequence
    expect(header1.equals(header2)).toBe(true)

    // note commitment
    header2.noteCommitment = Buffer.alloc(32, 'not  header')
    expect(header1.equals(header2)).toBe(false)
    header2.noteCommitment = header1.noteCommitment
    expect(header1.equals(header2)).toBe(true)

    // note size
    header2.noteSize = 7
    expect(header1.equals(header2)).toBe(false)
    header2.noteSize = header1.noteSize
    expect(header1.equals(header2)).toBe(true)

    // target
    header2.target = new Target(10)
    expect(header1.equals(header2)).toBe(false)
    header2.target = header1.target
    expect(header1.equals(header2)).toBe(true)

    // randomness
    header2.randomness = BigInt(19)
    expect(header1.equals(header2)).toBe(false)
    header2.randomness = header1.randomness
    expect(header1.equals(header2)).toBe(true)

    // timestamp
    header2.timestamp = new Date(1000)
    expect(header1.equals(header2)).toBe(false)
    header2.timestamp = header1.timestamp
    expect(header1.equals(header2)).toBe(true)

    // graffiti
    header2.graffiti = Buffer.alloc(32, 'a')
    expect(header1.equals(header2)).toBe(false)
    header2.graffiti = header1.graffiti
    expect(header1.equals(header2)).toBe(true)
  })
})

describe('BlockHeaderSerde', () => {
  const serde = BlockHeaderSerde

  it('serializes and deserializes a block header', () => {
    const header = new BlockHeader(
      5,
      Buffer.alloc(32),
      Buffer.alloc(32),
      Buffer.alloc(32, 'transactionRoot'),
      new Target(17),
      BigInt(25),
      new Date(1598467858637),
      GraffitiUtils.fromString('test'),
    )

    const serialized = serde.serialize(header)
    const deserialized = serde.deserialize(serialized)
    expect(header.equals(deserialized)).toBe(true)
  })

  it('checks block is later than', () => {
    const header1 = new BlockHeader(
      5,
      Buffer.alloc(32),
      Buffer.alloc(32),
      Buffer.alloc(32, 'transactionRoot'),
      new Target(0),
      BigInt(0),
      new Date(0),
      Buffer.alloc(32),
    )

    const serialized = serde.serialize(header1)
    const header2 = serde.deserialize(serialized)
    expect(isBlockLater(header1, header2)).toBe(false)

    header1.sequence = 6
    header2.sequence = 5
    expect(isBlockLater(header1, header2)).toBe(true)

    header1.sequence = 5
    header2.sequence = 5
    header1.hash = Buffer.alloc(32, 1)
    header2.hash = Buffer.alloc(32, 2)
    expect(isBlockLater(header1, header2)).toBe(true)
  })

  it('checks block is heavier than', () => {
    const header1 = new BlockHeader(
      5,
      Buffer.alloc(32),
      Buffer.alloc(32),
      Buffer.alloc(32, 'transactionRoot'),
      new Target(1),
      BigInt(0),
      new Date(0),
      Buffer.alloc(32),
    )

    const serialized = serde.serialize(header1)
    const header2 = serde.deserialize(serialized)
    expect(isBlockHeavier(header1, header2)).toBe(false)

    header1.work = BigInt(1)
    header2.work = BigInt(0)
    header1.sequence = 5
    header2.sequence = 5
    header1.target = new Target(100)
    header2.target = new Target(100)
    header1.hash = Buffer.alloc(32, 0)
    header1.hash = Buffer.alloc(32, 0)
    expect(isBlockHeavier(header1, header2)).toBe(true)

    header1.work = BigInt(0)
    header2.work = BigInt(0)
    header1.sequence = 6
    header2.sequence = 5
    header1.target = new Target(100)
    header2.target = new Target(100)
    header1.hash = Buffer.alloc(32, 0)
    header1.hash = Buffer.alloc(32, 0)
    expect(isBlockHeavier(header1, header2)).toBe(true)

    header1.work = BigInt(0)
    header2.work = BigInt(0)
    header1.sequence = 5
    header2.sequence = 5
    header1.target = new Target(100)
    header2.target = new Target(200)
    header1.hash = Buffer.alloc(32, 0)
    header1.hash = Buffer.alloc(32, 0)
    expect(isBlockHeavier(header1, header2)).toBe(true)

    header1.work = BigInt(0)
    header2.work = BigInt(0)
    header1.sequence = 5
    header2.sequence = 5
    header1.target = new Target(100)
    header2.target = new Target(100)
    header1.hash = Buffer.alloc(32, 0)
    header2.hash = Buffer.alloc(32, 1)
    expect(isBlockHeavier(header1, header2)).toBe(true)
  })
})
