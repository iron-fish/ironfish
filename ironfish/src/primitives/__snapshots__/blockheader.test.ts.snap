// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block Header Serde serializes and deserializes a block header 1`] = `
Object {
  "graffiti": "7465737400000000000000000000000000000000000000000000000000000000",
  "hash": "0500000000000000000000000000000000000000000000000000000000000000",
  "minersFee": "-1",
  "noteCommitment": Object {
    "commitment": "header",
    "size": 8,
  },
  "nullifierCommitment": Object {
    "commitment": "0000000000000000000000000000000000000000000000000000000000000000",
    "size": 3,
  },
  "previousBlockHash": "0000000000000000000000000000000000000000000000000000000000000000",
  "randomness": 25,
  "sequence": "5",
  "target": "17",
  "timestamp": 1598467858637,
  "work": "0",
}
`;
